posts/2023-10-05-Rust-1.73.0.md

```diff
@@ -0,0 +1,114 @@
+
+<div style="margin:1em"><pre><code>fn main() {
+    let file = "ferris.txt";
+    panic!("oh no! {file:?} not found!");
+}</code></pre>
+Output before Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at 'oh no! "ferris.txt" not found!', src/main.rs:3:5</code></pre>
+Output starting in Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at src/main.rs:3:5:
+oh no! "ferris.txt" not found!</code></pre></div>
+
+This is especially useful when the message is long, contains nested quotes, or spans multiple lines.
+
+Additionally, the panic messages produced by `assert_eq` and `assert_ne` have
+been modified, moving the custom message (the third argument)
+and removing some unnecessary punctuation, as shown below:
+
+<div style="margin:1em"><pre><code>fn main() {
+    assert_eq!("🦀", "🐟", "ferris is not a fish");
+}</code></pre>
+Output before Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at 'assertion failed: `(left == right)`
+ left: `"🦀"`,
+right: `"🐟"`: ferris is not a fish', src/main.rs:2:5</code></pre>
+Output starting in Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at src/main.rs:2:5:
+assertion `left == right` failed: ferris is not a fish
+ left: "🦀"
+right: "🐟"</code></pre></div>
+
+### Thread local initialization
+
+As proposed in [RFC 3184](https://github.com/rust-lang/rfcs/blob/master/text/3184-thread-local-cell-methods.md), `LocalKey<Cell<T>>` and `LocalKey<RefCell<T>>` can now be directly manipulated with `get()`, `set()`, `take()`, and `replace()` methods, rather than jumping through a `with(|inner| ...)` closure as needed for general `LocalKey` work. `LocalKey<T>` is the type of `thread_local!` statics.
+
+The new methods make common code more concise and avoid running the extra initialization code for the default value specified in `thread_local!` for new threads.
+
+```rust
+thread_local! {
+    static THINGS: Cell<Vec<i32>> = Cell::new(Vec::new());
+}
+
+fn f() {
+    // before:
+    THINGS.with(|i| i.set(vec![1, 2, 3]));
+    // now:
+    THINGS.set(vec![1, 2, 3]);
+
+    // ...
+
+    // before:
+    let v = THINGS.with(|i| i.take());
+    // now:
+    let v: Vec<i32> = THINGS.take();
+}
+```
+
+### Stabilized APIs
+
+- [Unsigned `{integer}::div_ceil`](https://doc.rust-lang.org/stable/std/primitive.u32.html#method.div_ceil)
+- [Unsigned `{integer}::next_multiple_of`](https://doc.rust-lang.org/stable/std/primitive.u32.html#method.next_multiple_of)
+- [Unsigned `{integer}::checked_next_multiple_of`](https://doc.rust-lang.org/stable/std/primitive.u32.html#method.checked_next_multiple_of)
+- [`std::ffi::FromBytesUntilNulError`](https://doc.rust-lang.org/stable/std/ffi/struct.FromBytesUntilNulError.html)
```

Comment:
Why is `FromBytesUntilNulError` on this list? It has been stable since 1.69 [[source](https://doc.rust-lang.org/stable/src/core/ffi/c_str.rs.html#155-157)] [[1.69 blog post](https://blog.rust-lang.org/2023/04/20/Rust-1.69.0.html#stabilized-apis)]. Was there something else that was supposed to be listed instead of this?

<img width="716" alt="image" src="https://github.com/rust-lang/blog.rust-lang.org/assets/5838512/58b4e870-88ce-4a20-a61d-8ba083feac0b">

Reply:
Not sure. @cuviper do you remember anything about why you added this?

Reply:
It was **only** in `core::ffi` before, forgotten to re-export from `std::ffi` [until now](rust-lang/rust/pull/113701). Compare:
* https://doc.rust-lang.org/1.72.0/std/ffi/#structs
* https://doc.rust-lang.org/1.73.0/std/ffi/#structs

Reply:
It was stabilized in 1.69 as `core::ffi::FromBytesUntilNulError` (notice the "core"), but was mistakenly not re-exported in std. 1.73 fixed that so that it is now available as `std::ffi::FromBytesUntilNulError`, and thus it is "new" at that location.

Reply:
Ohhhh